### PR TITLE
PR for dev branch

### PR DESCRIPTION
**Pull request by Aravindan07**

### CODE REVIEW

File: controllers/quiz.controller.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	try {
+		if (category === "random") {
+			const questions = await Quiz.find({}).select("-__v");
+			return res.status(200).json({ message: "Fetched successfully!", questions });
+		}
+		const questions = await Quiz.find({ category }).select("-__v");
+		res.status(200).json({ message: "Fetched successfully!", questions });
+	} catch (error) {
+		console.error(error);
+		res.status(500).json({ message: "Failed to fetch data" });
```
Comment (Jaynil1611) [inline controllers/quiz.controller.js:14]: You can use `next(error)` here as you are returning `500` status and leave it for error handler middleware to catch it
Apply the same for every controller function

File: controllers/quiz.controller.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	}
+};
+
+const postQuiz = async (req, res) => {
+	const { question, correctAnswer, options, category, difficulty } = req.body;
+	try {
+		const newQuiz = new Quiz({ question, options, correctAnswer, category, difficulty });
+		let result = await newQuiz.save();
```
Comment (Jaynil1611) [inline controllers/quiz.controller.js:22]: use `const` instead of `let`

File: controllers/score.controller.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+const Score = require("../models/score.model");
+const User = require("../models/user.model");
+const { getMaximumNumberInArray } = require("../utils/helper");
+
+const getUserDashboard = async (req, res) => {
+	const { userId } = req.params;
+	try {
+		const foundUserScore = await Score.findOne({ userId });
```
Comment (Jaynil1611) [inline controllers/score.controller.js:8]: Rename `foundUserScore` to `userScore`

File: controllers/score.controller.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		if (foundUserScore) {
+			return res.status(200).json({ item: foundUserScore });
+		}
+		return res.status(404).json({ message: "User not found!" });
```
Comment (Jaynil1611) [inline controllers/score.controller.js:12]: This error should be returned directly by the `authVerify` middleware and shouldn't reach here

File: controllers/score.controller.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+const Score = require("../models/score.model");
+const User = require("../models/user.model");
+const { getMaximumNumberInArray } = require("../utils/helper");
+
+const getUserDashboard = async (req, res) => {
+	const { userId } = req.params;
+	try {
+		const foundUserScore = await Score.findOne({ userId });
+		if (foundUserScore) {
+			return res.status(200).json({ item: foundUserScore });
```
Comment (Jaynil1611) [inline controllers/score.controller.js:10]: Keep the response structure same for `res.json` in `success` state as in send `.json({message:'', item:{}})`
As you have done in previous controller

File: controllers/score.controller.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		}
+		return res.status(404).json({ message: "User not found!" });
+	} catch (error) {
+		console.error(error);
+		return res.status(500).json({ message: "An error occurred" });
```
Comment (Jaynil1611) [inline controllers/score.controller.js:15]: Not needed here, let error middleware take care of this

File: controllers/score.controller.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+						: quiz
+				);
+				const newScore = await foundUserScore.save();
+				newScore.__v = undefined;
+				return res.status(201).json({ message: "Score added", item: newScore });
+			}
+			foundUserScore.quiz.push({ quizName, scores: [score], highScore: score });
+			const newUserScore = await foundUserScore.save();
+			newUserScore.__v = undefined;
+			return res.status(201).json({ message: "Score added", item: newUserScore });
```
Comment (Jaynil1611) [inline controllers/score.controller.js:41]: Try avoiding nested if statements, you can extract out the code in a separate function(s) and call them here
Make it easy to read for the user

File: controllers/user.controller.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+const validator = require("email-validator");
+const User = require("../models/user.model");
+
+const userRegister = async (req, res) => {
+	const { name, email, password } = req.body;
+
+	const checkValidation = validator.validate(email);
+
+	// Simple validation
+	if (!name || !email || !password) {
```
Comment (Jaynil1611) [inline controllers/user.controller.js:12]: Extract these conditions into a function named `validateUserDetails` so it makes it easy to read

File: controllers/user.controller.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
+const bcrypt = require("bcryptjs");
+const jwt = require("jsonwebtoken");
+const validator = require("email-validator");
+const User = require("../models/user.model");
+
+const userRegister = async (req, res) => {
```
Comment (Jaynil1611) [inline controllers/user.controller.js:6]: Change function name to verb+noun format -> `registerUser`

File: controllers/user.controller.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+	// Simple validation
+	if (!name || !email || !password) {
+		return res.status(400).json({ message: "Please Enter all fields" });
+	}
+	try {
+		if (checkValidation === false) {
+			return res.status(400).json({ message: "Entered email is not in the valid format" });
+		} else {
+			// check for existing user
```
Comment (Jaynil1611) [inline controllers/user.controller.js:19]: You are returning from the above statement so no need to use else condition here
Remove it

File: controllers/user.controller.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+			const user = await User.findOne({ email: email });
```
Comment (Jaynil1611) [inline controllers/user.controller.js:20]: You can change this to only `{email}`

File: controllers/user.controller.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+			if (user) {
+				return res.status(400).json({ message: "This user already exists" });
+			}
+			const newUser = new User({ name: name, email: email, password: password });
```
Comment (Jaynil1611) [inline controllers/user.controller.js:24]: You can change this to `{name, email, password}`

File: controllers/user.controller.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	} catch (error) {
+		console.error(error);
+		res.status(500).json({ message: "An error occurred" });
+	}
+};
+
+const loadUser = async (req, res) => {
+	try {
+		const user = await User.findById(req.user.id)
+			.select("-password -createdAt -updatedAt -__v")
```
Comment (Jaynil1611) [inline controllers/user.controller.js:99]: I really liked this `select` operator, I will also use this to remove unnecessary fields sent to UI

File: routes/score.router.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+const express = require("express");
+const router = express.Router();
+const { addScore, getUserDashboard } = require("../controllers/score.controller");
+const { checkAuth } = require("../middlewares/checkAuth");
+
+router.route("/:userId/get-user-dashboard").all(checkAuth).get(getUserDashboard);
```
Comment (Jaynil1611) [inline routes/score.router.js:6]: What does this `.all()` function do? Does it behave like middleware?

**Review by Jaynil1611 — CHANGES_REQUESTED** [review]
Hey, @Aravindan07 great work on the backend and UI if the quiz is awesome
The dashboard feature is great!

Really liked a few of the things which you have used!
Here are my few suggestions:

File: routes/score.router.js (hunk 14 above)
Comment (Aravindan07) [reply routes/score.router.js:6]: Yep, It is kind of a middleware which checks some condition on all the methods which are passing through this route.
For. eg., Here only the get method is present for this route if I add a post method here then that method also will be checked by this ```.all()``` middleware.

**Pull request merged**